DPPP/DPInfo.cc

```diff
@@ -70,35 +74,114 @@ void DPInfo::init(unsigned int ncorr, unsigned int startChan,
+    // Takes mean of middle elements if n is even; takes middle if odd.
+    ref_freq = 0.5 * (chan_freqs[(n - 1) / 2] + chan_freqs[n / 2]);
+  }
+
+  itsChanFreqs.clear();
+  itsChanWidths.clear();
+  itsResolutions.clear();
+  itsEffectiveBW.clear();
+
+  itsChanFreqs.push_back(std::move(chan_freqs));
+  itsChanWidths.push_back(std::move(chan_widths));
+  itsResolutions.push_back(std::move(resolutions));
+  itsEffectiveBW.push_back(std::move(effective_bw));
+
+  itsTotalBW = std::accumulate(itsEffectiveBW.front().begin(),
+                               itsEffectiveBW.front().end(), 0.0);
+  itsRefFreq = ref_freq;
+}
+
+void DPInfo::set(std::vector<std::vector<double>>&& chan_freqs,
+                 std::vector<std::vector<double>>&& chan_widths,
+                 std::vector<std::vector<double>>&& resolutions,
+                 std::vector<std::vector<double>>&& effective_bw,
+                 double ref_freq) {
+  if (resolutions.empty()) {
+    resolutions = chan_widths;
+  }
+  if (effective_bw.empty()) {
+    effective_bw = chan_widths;
+  }
+  if (chan_freqs.size() != nbaselines() || chan_widths.size() != nbaselines() ||
+      resolutions.size() != nbaselines() ||
+      effective_bw.size() != nbaselines()) {
+    throw Exception("Invalid baseline count while setting frequency info");
+  }
+
+  const double total_bw = std::accumulate(effective_bw.front().begin(),
+                                          effective_bw.front().end(), 0.0);
+  for (std::vector<double>& eff_bw_bl : effective_bw) {
+    if (std::accumulate(eff_bw_bl.begin(), eff_bw_bl.end(), 0.0) != total_bw) {
+      throw Exception("Total BW is not equal for all baselines");
+    }
+  }
+
+  if (ref_freq == 0) {
+    // Find the baseline with the most channels.
+    auto comp = [](const std::vector<double>& left,
+                   const std::vector<double>& right) {
+      return left.size() < right.size();
+    };
+    auto it = std::max_element(chan_freqs.begin(), chan_freqs.end(), comp);
+
+    std::size_t n = it->size();
     // Takes mean of middle elements if n is even; takes middle if odd.
-    itsRefFreq = 0.5 * (itsChanFreqs[(n - 1) / 2] + itsChanFreqs[n / 2]);
-  } else {
-    itsRefFreq = refFreq;
+    ref_freq = 0.5 * ((*it)[(n - 1) / 2] + (*it)[n / 2]);
+  }
+
+  itsChanFreqs = std::move(chan_freqs);
+  itsChanWidths = std::move(chan_widths);
+  itsResolutions = std::move(resolutions);
+  itsEffectiveBW = std::move(effective_bw);
+  itsTotalBW = total_bw;
+  itsRefFreq = ref_freq;
+}
+
+bool DPInfo::channelsAreRegular() const {
+  if (itsChanFreqs.empty()) {
+    return true;
+  }
+
+  // Check that all baselines have equal channel layouts.
+  for (std::size_t bl = 1; bl < itsChanFreqs.size(); ++bl) {
+    if (itsChanFreqs.front() != itsChanFreqs[bl] ||
+        itsChanWidths.front() != itsChanWidths[bl] ||
+        itsResolutions.front() != itsResolutions[bl] ||
+        itsEffectiveBW.front() != itsEffectiveBW[bl]) {
+      return false;
+    }
```

Comment:
@tammojan : Should we also apply the 1 kHz tolerance (see below) to these comparisons?